src/NHibernate.Test/Linq/ParameterTests.cs

```diff
@@ -0,0 +1,217 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text.RegularExpressions;
+using NHibernate.DomainModel.Northwind.Entities;
+using NUnit.Framework;
+
+namespace NHibernate.Test.Linq
+{
+	[TestFixture]
+	public class ParameterTests : LinqTestCase
+	{
+		[Test]
+		public void UsingArrayParameterTwice()
+		{
+			var ids = new[] {11008, 11019, 11039};
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids.Contains(o.OrderId)),
+				ids.Length);
+		}
+
+		[Test]
+		public void UsingTwoArrayParameters()
+		{
+			var ids = new[] {11008, 11019, 11039};
+			var ids2 = new[] {11008, 11019, 11039};
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids2.Contains(o.OrderId)),
+				ids.Length + ids2.Length);
+		}
+
+		[Test]
+		public void UsingListParameterTwice()
+		{
+			var ids = new List<int> {11008, 11019, 11039};
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids.Contains(o.OrderId)),
+				ids.Count);
+		}
+
+		[Test]
+		public void UsingTwoListParameters()
+		{
+			var ids = new List<int> {11008, 11019, 11039};
+			var ids2 = new List<int> {11008, 11019, 11039};
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids2.Contains(o.OrderId)),
```

Comment:
Just curious how will it handle `new[]{ value, 2}.Contains(o.OrderId) && new []{value, 3}.Contains(o.OrderId)`.

Reply:
With the current logic 4 parameters will be added. This is quite a special case as `new[]{ value, 2}` is treated as one parameter and expanded later when executing the query.

Reply:
Maybe cover it with test? And could you also add test for DML LINQ query.
Will this logic work for queries like (I'm actually not sure if it should be done for DML linq queries):
```C#
db.Customers.Where(c => c.CustomerId == "UNKNOWN").Update(x => new Customer {ContactName = value, CompanyName = value});
```

Reply:
> Maybe cover it with test?

Added.

> And could you also add test for DML LINQ query

I forgot about them. After some testing I found out that HQL does not support reusing parameters for DML queries, so now the new logic is not used for them anymore. Also, tests were added.